File: sample_files/golang_files/web_crypto_vulnerabilities.go
package main

/*
Go Web Service Cryptographic Vulnerabilities
Testing file for IBM Quantum Safe Explorer

WARNING: This code is intentionally insecure and should NOT be used in production!
*/

import (
	"crypto/md5"
	"crypto/rand"
	"crypto/rsa"
	"crypto/sha1"
	"crypto/sha256"
	"encoding/base64"
	"encoding/hex"
	"fmt"
	"math/big"
	mathrand "math/rand"
	"time"
)

// ============================================================================
// JWT VULNERABILITIES
// ============================================================================

// InsecureJWTHandler - Vulnerability: Weak JWT implementation
type InsecureJWTHandler struct {
	// Vulnerability: Weak secret key
	Secret string
}

// NewInsecureJWTHandler creates a handler with weak secret
func NewInsecureJWTHandler() *InsecureJWTHandler {
	return &InsecureJWTHandler{
		Secret: "weak_secret", // Vulnerability: Hardcoded weak secret
	}
}

// CreateJWTNoneAlgorithm - Vulnerability: JWT with 'none' algorithm
func (h *InsecureJWTHandler) CreateJWTNoneAlgorithm(userID int) string {
	header := base64.StdEncoding.EncodeToString([]byte(`{"alg":"none","typ":"JWT"}`))
	payload := base64.StdEncoding.EncodeToString([]byte(fmt.Sprintf(`{"user_id":%d}`, userID)))
	return fmt.Sprintf("%s.%s.", header, payload)
}

// CreateJWTWeakSecret - Vulnerability: JWT with weak secret
func (h *InsecureJWTHandler) CreateJWTWeakSecret(userID int) string {
	// Vulnerability: Using MD5 for HMAC
	data := fmt.Sprintf(`{"user_id":%d}`, userID)
	hash := md5.Sum([]byte(h.Secret + data))
	signature := base64.StdEncoding.EncodeToString(hash[:])
	return fmt.Sprintf("header.%s.%s", base64.StdEncoding.EncodeToString([]byte(data)), signature)
}

// CreateJWTRSAWeak - Vulnerability: JWT with RSA (quantum vulnerable)
func (h *InsecureJWTHandler) CreateJWTRSAWeak(userID int) (string, error) {
	// Vulnerability: Using weak RSA key
	privateKey, err := rsa.GenerateKey(rand.Reader, 1024)
	if err != nil {
		return "", err
	}

	data := []byte(fmt.Sprintf(`{"user_id":%d}`, userID))
	hashed := sha256.Sum256(data)
	signature, err := rsa.SignPKCS1v15(rand.Reader, privateKey, 0, hashed[:])
	if err != nil {
		return "", err
	}

	return base64.StdEncoding.EncodeToString(signature), nil
}

// ============================================================================
// SESSION MANAGEMENT VULNERABILITIES
// ============================================================================

// InsecureSessionManager - Vulnerability: Weak session management
type InsecureSessionManager struct {
	// Vulnerability: Hardcoded secret
	SecretKey string
}

// NewInsecureSessionManager creates a session manager with hardcoded secret
func NewInsecureSessionManager() *InsecureSessionManager {
	return &InsecureSessionManager{
		SecretKey: "hardcoded-session-secret", // Vulnerability
	}
}

// CreateSessionTokenWeak - Vulnerability: Weak session token generation
func (sm *InsecureSessionManager) CreateSessionTokenWeak(userID int) string {
	mathrand.Seed(int64(userID)) // Vulnerability: Predictable seed
	token := make([]byte, 32)
	for i := range token {
		token[i] = byte(mathrand.Intn(256))
	}
	return hex.EncodeToString(token)
}

// SignSessionDataMD5 - Vulnerability: Using MD5 for HMAC
func (sm *InsecureSessionManager) SignSessionDataMD5(data string) string {
	hash := md5.Sum([]byte(sm.SecretKey + data))
	return hex.EncodeToString(hash[:])
}

// SignSessionDataSHA1 - Vulnerability: Using SHA-1 for HMAC
func (sm *InsecureSessionManager) SignSessionDataSHA1(data string) string {
	hash := sha1.Sum([]byte(sm.SecretKey + data))
	return hex.EncodeToString(hash[:])
}

// ============================================================================
// API KEY GENERATION VULNERABILITIES
// ============================================================================

// InsecureAPIKeyGenerator - Vulnerability: Weak API key generation
type InsecureAPIKeyGenerator struct{}

// GenerateAPIKeyWeakRandom - Vulnerability: Using weak random for API keys
func (g *InsecureAPIKeyGenerator) GenerateAPIKeyWeakRandom() string {
	mathrand.Seed(time.Now().UnixNano()) // Vulnerability: Predictable
	key := make([]byte, 32)
	for i := range key {
		key[i] = byte(mathrand.Intn(256))
	}
	return hex.EncodeToString(key)
}

// GenerateAPIKeyPredictable - Vulnerability: Predictable API key based on user ID
func (g *InsecureAPIKeyGenerator) GenerateAPIKeyPredictable(userID int) string {
	data := fmt.Sprintf("api_key_%d", userID)
	hash := md5.Sum([]byte(data))
	return hex.EncodeToString(hash[:])
}

// GenerateAPIKeySequential - Vulnerability: Sequential API keys
func (g *InsecureAPIKeyGenerator) GenerateAPIKeySequential(counter int) string {
	return fmt.Sprintf("API-%08d", counter)
}

// ============================================================================
// PASSWORD HASHING VULNERABILITIES
// ============================================================================

// InsecurePasswordHasher - Vulnerability: Weak password hashing
type InsecurePasswordHasher struct{}

// HashPasswordMD5 - Vulnerability: MD5 for password hashing
func (ph *InsecurePasswordHasher) HashPasswordMD5(password string) string {
	hash := md5.Sum([]byte(password))
	return hex.EncodeToString(hash[:])
}

// HashPasswordSHA1 - Vulnerability: SHA-1 for password hashing
func (ph *InsecurePasswordHasher) HashPasswordSHA1(password string) string {
	hash := sha1.Sum([]byte(password))
	return hex.EncodeToString(hash[:])
}

// HashPasswordMD5WithSalt - Vulnerability: MD5 with weak salt
func (ph *InsecurePasswordHasher) HashPasswordMD5WithSalt(password string) string {
	mathrand.Seed(time.Now().UnixNano())
	salt := fmt.Sprintf("%d", mathrand.Intn(10000)) // Weak salt
	hash := md5.Sum([]byte(salt + password))
	return fmt.Sprintf("%s:%s", salt, hex.EncodeToString(hash[:]))
}

// VerifyPasswordTimingAttack - Vulnerability: Timing attack in password verification
func (ph *InsecurePasswordHasher) VerifyPasswordTimingAttack(password, hashed string) bool {
	return ph.HashPasswordMD5(password) == hashed // Not constant-time
}

// ============================================================================
// PASSWORD RESET TOKEN VULNERABILITIES
// ============================================================================

// InsecurePasswordResetHandler - Vulnerability: Weak password reset tokens
type InsecurePasswordResetHandler struct{}

// GenerateResetTokenWeak - Vulnerability: Predictable reset token
func (h *InsecurePasswordResetHandler) GenerateResetTokenWeak(email string) string {
	timestamp := time.Now().Unix()
	data := fmt.Sprintf("%s%d", email, timestamp)
	hash := md5.Sum([]byte(data))
	return hex.EncodeToString(hash[:])
}

// GenerateResetTokenNoExpiry - Vulnerability: Reset token without expiration
func (h *InsecurePasswordResetHandler) GenerateResetTokenNoExpiry(userID int) string {
	mathrand.Seed(int64(userID))
	token := make([]byte, 6)
	for i := range token {
		token[i] = byte('0' + mathrand.Intn(10))
	}
	return string(token)
}

// VerifyResetTokenTiming - Vulnerability: Timing attack in token verification
func (h *InsecurePasswordResetHandler) VerifyResetTokenTiming(token, expected string) bool {
	return token == expected // Not constant-time comparison
}

// ============================================================================
// TWO-FACTOR AUTHENTICATION VULNERABILITIES
// ============================================================================

// Insecure2FAHandler - Vulnerability: Weak 2FA implementation
type Insecure2FAHandler struct{}

// GenerateTOTPSecretWeak - Vulnerability: Weak TOTP secret generation
func (h *Insecure2FAHandler) GenerateTOTPSecretWeak() string {
	mathrand.Seed(42) // Vulnerability: Fixed seed
	secret := make([]byte, 10)
	for i := range secret {
		secret[i] = byte(mathrand.Intn(256))
	}
	return base64.StdEncoding.EncodeToString(secret)
}

// GenerateBackupCodesWeak - Vulnerability: Weak backup code generation
func (h *Insecure2FAHandler) GenerateBackupCodesWeak(count int) []string {
	mathrand.Seed(time.Now().UnixNano())
	codes := make([]string, count)
	for i := 0; i < count; i++ {
		code := make([]byte, 8)
		for j := range code {
			code[j] = byte('0' + mathrand.Intn(10))
		}
		codes[i] = string(code)
	}
	return codes
}

// Verify2FACodeTiming - Vulnerability: Timing attack in 2FA verification
func (h *Insecure2FAHandler) Verify2FACodeTiming(code, expected string) bool {
	return code == expected // Not constant-time
}

// ============================================================================
// OAUTH/OPENID VULNERABILITIES
// ============================================================================

// InsecureOAuthHandler - Vulnerability: Weak OAuth implementation
type InsecureOAuthHandler struct{}

// GenerateStateParameterWeak - Vulnerability: Weak OAuth state parameter
func (h *InsecureOAuthHandler) GenerateStateParameterWeak() string {
	mathrand.Seed(time.Now().UnixNano())
	state := make([]byte, 16)
	for i := range state {
		state[i] = byte('0' + mathrand.Intn(10))
	}
	return string(state)
}

// GenerateClientSecretWeak - Vulnerability: Predictable client secret
func (h *InsecureOAuthHandler) GenerateClientSecretWeak(clientID string) string {
	data := fmt.Sprintf("client_%s", clientID)
	hash := md5.Sum([]byte(data))
	return hex.EncodeToString(hash[:])
}

// GenerateAuthorizationCodeWeak - Vulnerability: Weak authorization code
func (h *InsecureOAuthHandler) GenerateAuthorizationCodeWeak(userID int) string {
	mathrand.Seed(int64(userID))
	code := make([]byte, 32)
	for i := range code {
		code[i] = byte(mathrand.Intn(256))
	}
	return hex.EncodeToString(code)
}

// ============================================================================
// COOKIE SECURITY VULNERABILITIES
// ============================================================================

// InsecureCookieHandler - Vulnerability: Insecure cookie handling
type InsecureCookieHandler struct {
	// Vulnerability: Hardcoded encryption key
	EncryptionKey []byte
}

// NewInsecureCookieHandler creates a handler with hardcoded key
func NewInsecureCookieHandler() *InsecureCookieHandler {
	return &InsecureCookieHandler{
		EncryptionKey: []byte("CookieKey1234567"), // Vulnerability
	}
}

// EncryptCookieWeak - Vulnerability: Weak cookie encryption
func (h *InsecureCookieHandler) EncryptCookieWeak(value string) string {
	// Vulnerability: Simple XOR (extremely weak)
	encrypted := make([]byte, len(value))
	for i := 0; i < len(value); i++ {
		encrypted[i] = value[i] ^ h.EncryptionKey[i%len(h.EncryptionKey)]
	}
	return base64.StdEncoding.EncodeToString(encrypted)
}

// SignCookieMD5 - Vulnerability: Using MD5 for cookie signature
func (h *InsecureCookieHandler) SignCookieMD5(value string) string {
	hash := md5.Sum([]byte(string(h.EncryptionKey) + value))
	return hex.EncodeToString(hash[:])
}

// ============================================================================
// CERTIFICATE PINNING VULNERABILITIES
// ============================================================================

// InsecureCertificateHandler - Vulnerability: Weak certificate handling
type InsecureCertificateHandler struct{}

// GenerateWeakCertificateFingerprint - Vulnerability: Using MD5 for cert fingerprint
func (h *InsecureCertificateHandler) GenerateWeakCertificateFingerprint(certData []byte) string {
	hash := md5.Sum(certData)
	return hex.EncodeToString(hash[:])
}

// GenerateSHA1CertificateFingerprint - Vulnerability: Using SHA-1 for cert fingerprint
func (h *InsecureCertificateHandler) GenerateSHA1CertificateFingerprint(certData []byte) string {
	hash := sha1.Sum(certData)
	return hex.EncodeToString(hash[:])
}

// ============================================================================
// RATE LIMITING TOKEN VULNERABILITIES
// ============================================================================

// InsecureRateLimiter - Vulnerability: Weak rate limiting tokens
type InsecureRateLimiter struct{}

// GenerateRateLimitTokenWeak - Vulnerability: Predictable rate limit token
func (rl *InsecureRateLimiter) GenerateRateLimitTokenWeak(clientIP string) string {
	timestamp := time.Now().Unix()
	data := fmt.Sprintf("%s:%d", clientIP, timestamp)
	hash := md5.Sum([]byte(data))
	return hex.EncodeToString(hash[:])
}

// ============================================================================
// MAIN DEMONSTRATION
// ============================================================================

func main() {
	fmt.Println("======================================================================")
	fmt.Println("WEB SERVICE CRYPTO VULNERABILITIES - IBM Quantum Safe Explorer")
	fmt.Println("======================================================================")

	// JWT vulnerabilities
	fmt.Println("\n[1] JWT VULNERABILITIES:")
	jwtHandler := NewInsecureJWTHandler()
	fmt.Printf("   JWT with 'none' algorithm: %s\n", jwtHandler.CreateJWTNoneAlgorithm(12345))
	fmt.Printf("   JWT with weak secret: %s\n", jwtHandler.CreateJWTWeakSecret(12345))

	// Session management vulnerabilities
	fmt.Println("\n[2] SESSION MANAGEMENT VULNERABILITIES:")
	sessionMgr := NewInsecureSessionManager()
	fmt.Printf("   Weak Session Token: %s\n", sessionMgr.CreateSessionTokenWeak(12345))
	fmt.Printf("   MD5 HMAC: %s\n", sessionMgr.SignSessionDataMD5("session_data"))

	// API key vulnerabilities
	fmt.Println("\n[3] API KEY GENERATION VULNERABILITIES:")
	apiGen := &InsecureAPIKeyGenerator{}
	fmt.Printf("   Weak Random API Key: %s\n", apiGen.GenerateAPIKeyWeakRandom())
	fmt.Printf("   Predictable API Key: %s\n", apiGen.GenerateAPIKeyPredictable(12345))

	// Password hashing vulnerabilities
	fmt.Println("\n[4] PASSWORD HASHING VULNERABILITIES:")
	pwHasher := &InsecurePasswordHasher{}
	fmt.Printf("   MD5 Password Hash: %s\n", pwHasher.HashPasswordMD5("password123"))
	fmt.Printf("   SHA-1 Password Hash: %s\n", pwHasher.HashPasswordSHA1("password123"))

	// Password reset vulnerabilities
	fmt.Println("\n[5] PASSWORD RESET TOKEN VULNERABILITIES:")
	resetHandler := &InsecurePasswordResetHandler{}
	fmt.Printf("   Weak Reset Token: %s\n", resetHandler.GenerateResetTokenWeak("user@example.com"))

	// 2FA vulnerabilities
	fmt.Println("\n[6] TWO-FACTOR AUTHENTICATION VULNERABILITIES:")
	twoFAHandler := &Insecure2FAHandler{}
	fmt.Printf("   Weak TOTP Secret: %s\n", twoFAHandler.GenerateTOTPSecretWeak())
	fmt.Printf("   Weak Backup Codes: %v\n", twoFAHandler.GenerateBackupCodesWeak(5))

	// OAuth vulnerabilities
	fmt.Println("\n[7] OAUTH VULNERABILITIES:")
	oauthHandler := &InsecureOAuthHandler{}
	fmt.Printf("   Weak State Parameter: %s\n", oauthHandler.GenerateStateParameterWeak())
	fmt.Printf("   Weak Client Secret: %s\n", oauthHandler.GenerateClientSecretWeak("client123"))

	// Cookie vulnerabilities
	fmt.Println("\n[8] COOKIE SECURITY VULNERABILITIES:")
	cookieHandler := NewInsecureCookieHandler()
	fmt.Printf("   Weak Cookie Encryption: %s\n", cookieHandler.EncryptCookieWeak("session_data"))
	fmt.Printf("   MD5 Cookie Signature: %s\n", cookieHandler.SignCookieMD5("session_data"))

	fmt.Println("\n======================================================================")
	fmt.Println("All web service vulnerability tests completed!")
	fmt.Println("This file should trigger multiple findings in Quantum Safe Explorer!")
	fmt.Println("======================================================================")
}

// Made with Bob

File: sample_files/csharp_files/WebCryptoVulnerabilities.cs
using System;
using System.Text;
using System.Security.Cryptography;
using System.Web;
using System.IdentityModel.Tokens.Jwt;
using Microsoft.IdentityModel.Tokens;

/*
 * C# ASP.NET Web Cryptographic Vulnerabilities
 * Testing file for IBM Quantum Safe Explorer
 * 
 * WARNING: This code is intentionally insecure and should NOT be used in production!
 */

namespace WebCryptoVulnerabilities
{
    // ============================================================================
    // JWT VULNERABILITIES
    // ============================================================================

    public class InsecureJWTHandler
    {
        // Vulnerability: Weak secret key
        private const string WeakSecret = "weak_secret";

        /// <summary>
        /// Vulnerability: JWT with weak secret
        /// </summary>
        public string CreateJWTWeakSecret(int userId)
        {
            var tokenHandler = new JwtSecurityTokenHandler();
            var key = Encoding.ASCII.GetBytes(WeakSecret);
            var tokenDescriptor = new SecurityTokenDescriptor
            {
                Subject = new System.Security.Claims.ClaimsIdentity(new[]
                {
                    new System.Security.Claims.Claim("user_id", userId.ToString())
                }),
                Expires = DateTime.UtcNow.AddHours(24),
                SigningCredentials = new SigningCredentials(
                    new SymmetricSecurityKey(key),
                    SecurityAlgorithms.HmacSha256Signature)
            };
            var token = tokenHandler.CreateToken(tokenDescriptor);
            return tokenHandler.WriteToken(token);
        }

        /// <summary>
        /// Vulnerability: JWT without expiration
        /// </summary>
        public string CreateJWTNoExpiration(int userId)
        {
            var tokenHandler = new JwtSecurityTokenHandler();
            var key = Encoding.ASCII.GetBytes(WeakSecret);
            var tokenDescriptor = new SecurityTokenDescriptor
            {
                Subject = new System.Security.Claims.ClaimsIdentity(new[]
                {
                    new System.Security.Claims.Claim("user_id", userId.ToString())
                }),
                // Vulnerability: No expiration set
                SigningCredentials = new SigningCredentials(
                    new SymmetricSecurityKey(key),
                    SecurityAlgorithms.HmacSha256Signature)
            };
            var token = tokenHandler.CreateToken(tokenDescriptor);
            return tokenHandler.WriteToken(token);
        }

        /// <summary>
        /// Vulnerability: JWT with RSA (quantum vulnerable)
        /// </summary>
        public string CreateJWTRSAWeak(int userId)
        {
            using (RSACryptoServiceProvider rsa = new RSACryptoServiceProvider(1024))
            {
                var tokenHandler = new JwtSecurityTokenHandler();
                var tokenDescriptor = new SecurityTokenDescriptor
                {
                    Subject = new System.Security.Claims.ClaimsIdentity(new[]
                    {
                        new System.Security.Claims.Claim("user_id", userId.ToString())
                    }),
                    Expires = DateTime.UtcNow.AddHours(24),
                    SigningCredentials = new SigningCredentials(
                        new RsaSecurityKey(rsa),
                        SecurityAlgorithms.RsaSha256)
                };
                var token = tokenHandler.CreateToken(tokenDescriptor);
                return tokenHandler.WriteToken(token);
            }
        }
    }

    // ============================================================================
    // SESSION MANAGEMENT VULNERABILITIES
    // ============================================================================

    public class InsecureSessionManager
    {
        // Vulnerability: Hardcoded secret
        private const string SecretKey = "hardcoded-session-secret";

        /// <summary>
        /// Vulnerability: Weak session token generation
        /// </summary>
        public string CreateSessionTokenWeak(int userId)
        {
            Random random = new Random(userId); // Vulnerability: Predictable seed
            byte[] token = new byte[32];
            random.NextBytes(token);
            return Convert.ToBase64String(token);
        }

        /// <summary>
        /// Vulnerability: Using MD5 for HMAC
        /// </summary>
        public string SignSessionDataMD5(string data)
        {
            using (HMACMD5 hmac = new HMACMD5(Encoding.UTF8.GetBytes(SecretKey)))
            {
                byte[] hash = hmac.ComputeHash(Encoding.UTF8.GetBytes(data));
                return Convert.ToBase64String(hash);
            }
        }

        /// <summary>
        /// Vulnerability: Using SHA-1 for HMAC
        /// </summary>
        public string SignSessionDataSHA1(string data)
        {
            using (HMACSHA1 hmac = new HMACSHA1(Encoding.UTF8.GetBytes(SecretKey)))
            {
                byte[] hash = hmac.ComputeHash(Encoding.UTF8.GetBytes(data));
                return Convert.ToBase64String(hash);
            }
        }
    }

    // ============================================================================
    // API KEY GENERATION VULNERABILITIES
    // ============================================================================

    public class InsecureAPIKeyGenerator
    {
        /// <summary>
        /// Vulnerability: Using weak random for API keys
        /// </summary>
        public string GenerateAPIKeyWeakRandom()
        {
            Random random = new Random((int)DateTime.Now.Ticks); // Vulnerability: Predictable
            byte[] key = new byte[32];
            random.NextBytes(key);
            return Convert.ToBase64String(key);
        }

        /// <summary>
        /// Vulnerability: Predictable API key based on user ID
        /// </summary>
        public string GenerateAPIKeyPredictable(int userId)
        {
            string data = $"api_key_{userId}";
            using (MD5 md5 = MD5.Create())
            {
                byte[] hash = md5.ComputeHash(Encoding.UTF8.GetBytes(data));
                return BitConverter.ToString(hash).Replace("-", "").ToLower();
            }
        }

        /// <summary>
        /// Vulnerability: Sequential API keys
        /// </summary>
        public string GenerateAPIKeySequential(int counter)
        {
            return $"API-{counter:D8}";
        }
    }

    // ============================================================================
    // PASSWORD HASHING VULNERABILITIES
    // ============================================================================

    public class InsecurePasswordHasher
    {
        /// <summary>
        /// Vulnerability: MD5 for password hashing
        /// </summary>
        public string HashPasswordMD5(string password)
        {
            using (MD5 md5 = MD5.Create())
            {
                byte[] hash = md5.ComputeHash(Encoding.UTF8.GetBytes(password));
                return Convert.ToBase64String(hash);
            }
        }

        /// <summary>
        /// Vulnerability: SHA-1 for password hashing
        /// </summary>
        public string HashPasswordSHA1(string password)
        {
            using (SHA1 sha1 = SHA1.Create())
            {
                byte[] hash = sha1.ComputeHash(Encoding.UTF8.GetBytes(password));
                return Convert.ToBase64String(hash);
            }
        }

        /// <summary>
        /// Vulnerability: MD5 with weak salt
        /// </summary>
        public string HashPasswordMD5WithSalt(string password)
        {
            Random random = new Random((int)DateTime.Now.Ticks);
            string salt = random.Next(10000).ToString(); // Weak salt
            using (MD5 md5 = MD5.Create())
            {
                byte[] hash = md5.ComputeHash(Encoding.UTF8.GetBytes(salt + password));
                return $"{salt}:{Convert.ToBase64String(hash)}";
            }
        }

        /// <summary>
        /// Vulnerability: Timing attack in password verification
        /// </summary>
        public bool VerifyPasswordTimingAttack(string password, string hashed)
        {
            return HashPasswordMD5(password) == hashed; // Not constant-time
        }
    }

    // ============================================================================
    // PASSWORD RESET TOKEN VULNERABILITIES
    // ============================================================================

    public class InsecurePasswordResetHandler
    {
        /// <summary>
        /// Vulnerability: Predictable reset token
        /// </summary>
        public string GenerateResetTokenWeak(string email)
        {
            long timestamp = DateTime.Now.Ticks;
            string data = $"{email}{timestamp}";
            using (MD5 md5 = MD5.Create())
            {
                byte[] hash = md5.ComputeHash(Encoding.UTF8.GetBytes(data));
                return BitConverter.ToString(hash).Replace("-", "").ToLower();
            }
        }

        /// <summary>
        /// Vulnerability: Reset token without expiration
        /// </summary>
        public string GenerateResetTokenNoExpiry(int userId)
        {
            Random random = new Random(userId);
            StringBuilder token = new StringBuilder();
            for (int i = 0; i < 6; i++)
            {
                token.Append(random.Next(10));
            }
            return token.ToString();
        }

        /// <summary>
        /// Vulnerability: Timing attack in token verification
        /// </summary>
        public bool VerifyResetTokenTiming(string token, string expected)
        {
            return token == expected; // Not constant-time comparison
        }
    }

    // ============================================================================
    // TWO-FACTOR AUTHENTICATION VULNERABILITIES
    // ============================================================================

    public class Insecure2FAHandler
    {
        /// <summary>
        /// Vulnerability: Weak TOTP secret generation
        /// </summary>
        public string GenerateTOTPSecretWeak()
        {
            Random random = new Random(42); // Vulnerability: Fixed seed
            byte[] secret = new byte[10];
            random.NextBytes(secret);
            return Convert.ToBase64String(secret);
        }

        /// <summary>
        /// Vulnerability: Weak backup code generation
        /// </summary>
        public string[] GenerateBackupCodesWeak(int count)
        {
            Random random = new Random((int)DateTime.Now.Ticks);
            string[] codes = new string[count];
            for (int i = 0; i < count; i++)
            {
                StringBuilder code = new StringBuilder();
                for (int j = 0; j < 8; j++)
                {
                    code.Append(random.Next(10));
                }
                codes[i] = code.ToString();
            }
            return codes;
        }

        /// <summary>
        /// Vulnerability: Timing attack in 2FA verification
        /// </summary>
        public bool Verify2FACodeTiming(string code, string expected)
        {
            return code == expected; // Not constant-time
        }
    }

    // ============================================================================
    // OAUTH/OPENID VULNERABILITIES
    // ============================================================================

    public class InsecureOAuthHandler
    {
        /// <summary>
        /// Vulnerability: Weak OAuth state parameter
        /// </summary>
        public string GenerateStateParameterWeak()
        {
            Random random = new Random((int)DateTime.Now.Ticks);
            StringBuilder state = new StringBuilder();
            for (int i = 0; i < 16; i++)
            {
                state.Append(random.Next(10));
            }
            return state.ToString();
        }

        /// <summary>
        /// Vulnerability: Predictable client secret
        /// </summary>
        public string GenerateClientSecretWeak(string clientId)
        {
            string data = $"client_{clientId}";
            using (MD5 md5 = MD5.Create())
            {
                byte[] hash = md5.ComputeHash(Encoding.UTF8.GetBytes(data));
                return BitConverter.ToString(hash).Replace("-", "").ToLower();
            }
        }

        /// <summary>
        /// Vulnerability: Weak authorization code
        /// </summary>
        public string GenerateAuthorizationCodeWeak(int userId)
        {
            Random random = new Random(userId);
            byte[] code = new byte[32];
            random.NextBytes(code);
            return Convert.ToBase64String(code);
        }
    }

    // ============================================================================
    // COOKIE SECURITY VULNERABILITIES
    // ============================================================================

    public class InsecureCookieHandler
    {
        // Vulnerability: Hardcoded encryption key
        private static readonly byte[] EncryptionKey = Encoding.UTF8.GetBytes("CookieKey1234567");

        /// <summary>
        /// Vulnerability: Weak cookie encryption
        /// </summary>
        public string EncryptCookieWeak(string value)
        {
            // Vulnerability: Simple XOR (extremely weak)
            byte[] valueBytes = Encoding.UTF8.GetBytes(value);
            byte[] encrypted = new byte[valueBytes.Length];
            for (int i = 0; i < valueBytes.Length; i++)
            {
                encrypted[i] = (byte)(valueBytes[i] ^ EncryptionKey[i % EncryptionKey.Length]);
            }
            return Convert.ToBase64String(encrypted);
        }

        /// <summary>
        /// Vulnerability: Using MD5 for cookie signature
        /// </summary>
        public string SignCookieMD5(string value)
        {
            using (MD5 md5 = MD5.Create())
            {
                byte[] data = Encoding.UTF8.GetBytes(Encoding.UTF8.GetString(EncryptionKey) + value);
                byte[] hash = md5.ComputeHash(data);
                return BitConverter.ToString(hash).Replace("-", "").ToLower();
            }
        }

        /// <summary>
        /// Vulnerability: Cookie without Secure flag
        /// </summary>
        public HttpCookie CreateInsecureCookie(string name, string value)
        {
            HttpCookie cookie = new HttpCookie(name, value);
            cookie.HttpOnly = false; // Vulnerable to XSS
            cookie.Secure = false;   // Can be transmitted over HTTP
            return cookie;
        }
    }

    // ============================================================================
    // VIEW STATE VULNERABILITIES
    // ============================================================================

    public class InsecureViewStateHandler
    {
        /// <summary>
        /// Vulnerability: ViewState with weak MAC
        /// </summary>
        public string GenerateViewStateMACWeak(string viewStateData)
        {
            using (MD5 md5 = MD5.Create())
            {
                byte[] hash = md5.ComputeHash(Encoding.UTF8.GetBytes(viewStateData));
                return Convert.ToBase64String(hash);
            }
        }
    }

    // ============================================================================
    // MAIN DEMONSTRATION
    // ============================================================================

    public class Program
    {
        public static void Main(string[] args)
        {
            Console.WriteLine("======================================================================");
            Console.WriteLine("WEB CRYPTO VULNERABILITIES - IBM Quantum Safe Explorer");
            Console.WriteLine("======================================================================");

            // JWT vulnerabilities
            Console.WriteLine("\n[1] JWT VULNERABILITIES:");
            InsecureJWTHandler jwtHandler = new InsecureJWTHandler();
            Console.WriteLine($"   JWT with weak secret: {jwtHandler.CreateJWTWeakSecret(12345)}");

            // Session management vulnerabilities
            Console.WriteLine("\n[2] SESSION MANAGEMENT VULNERABILITIES:");
            InsecureSessionManager sessionMgr = new InsecureSessionManager();
            Console.WriteLine($"   Weak Session Token: {sessionMgr.CreateSessionTokenWeak(12345)}");
            Console.WriteLine($"   MD5 HMAC: {sessionMgr.SignSessionDataMD5("session_data")}");

            // API key vulnerabilities
            Console.WriteLine("\n[3] API KEY GENERATION VULNERABILITIES:");
            InsecureAPIKeyGenerator apiGen = new InsecureAPIKeyGenerator();
            Console.WriteLine($"   Weak Random API Key: {apiGen.GenerateAPIKeyWeakRandom()}");
            Console.WriteLine($"   Predictable API Key: {apiGen.GenerateAPIKeyPredictable(12345)}");

            // Password hashing vulnerabilities
            Console.WriteLine("\n[4] PASSWORD HASHING VULNERABILITIES:");
            InsecurePasswordHasher pwHasher = new InsecurePasswordHasher();
            Console.WriteLine($"   MD5 Password Hash: {pwHasher.HashPasswordMD5("password123")}");
            Console.WriteLine($"   SHA-1 Password Hash: {pwHasher.HashPasswordSHA1("password123")}");

            // Password reset vulnerabilities
            Console.WriteLine("\n[5] PASSWORD RESET TOKEN VULNERABILITIES:");
            InsecurePasswordResetHandler resetHandler = new InsecurePasswordResetHandler();
            Console.WriteLine($"   Weak Reset Token: {resetHandler.GenerateResetTokenWeak("user@example.com")}");

            // 2FA vulnerabilities
            Console.WriteLine("\n[6] TWO-FACTOR AUTHENTICATION VULNERABILITIES:");
            Insecure2FAHandler twoFAHandler = new Insecure2FAHandler();
            Console.WriteLine($"   Weak TOTP Secret: {twoFAHandler.GenerateTOTPSecretWeak()}");
            Console.WriteLine($"   Weak Backup Codes: {string.Join(", ", twoFAHandler.GenerateBackupCodesWeak(5))}");

            // OAuth vulnerabilities
            Console.WriteLine("\n[7] OAUTH VULNERABILITIES:");
            InsecureOAuthHandler oauthHandler = new InsecureOAuthHandler();
            Console.WriteLine($"   Weak State Parameter: {oauthHandler.GenerateStateParameterWeak()}");
            Console.WriteLine($"   Weak Client Secret: {oauthHandler.GenerateClientSecretWeak("client123")}");

            // Cookie vulnerabilities
            Console.WriteLine("\n[8] COOKIE SECURITY VULNERABILITIES:");
            InsecureCookieHandler cookieHandler = new InsecureCookieHandler();
            Console.WriteLine($"   Weak Cookie Encryption: {cookieHandler.EncryptCookieWeak("session_data")}");
            Console.WriteLine($"   MD5 Cookie Signature: {cookieHandler.SignCookieMD5("session_data")}");

            Console.WriteLine("\n======================================================================");
            Console.WriteLine("All web vulnerability tests completed!");
            Console.WriteLine("This file should trigger multiple findings in Quantum Safe Explorer!");
            Console.WriteLine("======================================================================");
        }
    }
}

// Made with Bob